### PR TITLE
Feature: Grant followers modal

### PR DESCRIPTION
**Pull request by greg-adams**
## Description
- Adds modal for displaying grant followers
- Updates pagination for grant followers to allow for sorting by specific fields (created_at vs user name).

## Screenshots / Demo Video
Screenshot from 2024-09-13 14-14-05

### Automated and Unit Tests
- [x] Added Unit tests

### Manual tests for Reviewer
- [ ] Added steps to test feature/functionality manually

## Checklist
- [x] Provided ticket and description
- [x] Provided screenshots/demo
- [ ] Provided testing information
- [x] Provided adequate test coverage for all new code
- [x] Added PR reviewers

### CODE REVIEW

**Review by TylerHendrickson — CHANGES_REQUESTED** [review]
@greg-adams I have two visual formatting suggestions (see comments in `GrantFollowers.vue`) for you to consider. 

A bit more broadly, I don't think we need the changes to pagination behavior that this PR introduces. For clarification:
- The intended ordering behavior for the modal's list is most-recently-followed first (latest follower at the top, first follower at the bottom), not alphabetical.
- I'd like to generally move the API towards cursor pagination rather than traditional offset-based pagination.

(As always, happy to chat about this further if you have questions or think I missed something.)

----

Additional context to elaborate on the preference for cursor-based pagination, in case you're interested – feel free to disregard if not :) ...

With cursor-based pagination, the database returns a cursor (or "checkpoint") value pointing to the final item in the result set, which is then used as an "exclusive start" filter on a subsequent query. This is in contrast to offset-based pagination, where the query specifies a certain number of results to skip, based on a running total result count. As long as clients treat the cursor value (`pagination.from` in our case) as an arbitrary string, it allows us to more easily implement a different checkpoint mechanism for a particular retrieval scenario (e.g. maybe we return `from: grant_followers.created_at` instead of `from: grant_followers.id`).

One advantage of cursor pagination is that it's more accurate – offset-based pagination has a tendency to either skip or return duplicate items when the overall result set is being modified between queries. For example, in a scenario where an array of `Alice, Bob, Charlie, Elise, Frank, George` is being paged 1 result at a time, if the query returns `Frank` and then `Dan` is inserted between `Charlie` and `Elise`, an offset-based paginated request would return `Frank` a second time. A cursor-based paginated request would ensure that `George` (or another value alphabetically greater than `Frank`) always comes next.

Another advantage is that they tend to be more compatible with non-SQL backends, so it offers a certain amount of future-proofing in case we decide to move to an ordered result cache or other kind of precomputed index.

Note: Although we don't currently have a use-case for ordering followers alphabetically, it is possible to implement cursor-based pagination in a way that supports multiple ordering schemes – the `paginate.from` value would just need to be set to the `ORDER BY` column value of the last item in the page. We would also need to support an `afterName` parameter in `getFollowersForGrant()`, or else refactor to a more generic `startAfter` parameter.

**Comment by greg-adams** [comment]
Ok gotcha, this makes sense, thanks for the background. The story AC requests sorting alphabetically by name, if we don't need this change I can return to ordering by creation. While making this change I added support on the response to identify a final result set (`next`) to avoid the user clicking "Show More" an extra time at the end, do you have a preference for supporting this in cursor-based approach? e.g. `paginateBefore`

File: packages/client/src/components/Modals/GrantFollowers.vue [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+          let dateFollowedText = '';
+          if (createdDate.isSame(moment(), 'day')) {
+            dateFollowedText = 'Today';
+          } else if (createdDate.isAfter(moment().subtract(7, 'days'))) {
+            const dayCount = moment().diff(createdDate, 'days');
+            dateFollowedText = `${dayCount} ${dayCount === 1 ? 'day' : 'days'} ago`;
+          } else {
+            dateFollowedText = createdDate.format('MMMM D');
+          }
```
[the remark this answers is not in the document]
Comment (greg-adams) [reply packages/client/src/components/Modals/GrantFollowers.vue:137]: No problem, I can re-work to use luxon

**Comment by TylerHendrickson** [comment]
@greg-adams 

> The story AC requests sorting alphabetically by name, if we don't need this change I can return to ordering by creation. 

Yeah, I just noticed that; we'd [discussed in the comments](https://github.com/usdigitalresponse/usdr-gost/issues/2960#issuecomment-2293722197) of the parent issue (#2960) and decided to go with with most-recent-first instead, but it seems the DoD on #3407 was never updated to reflect that decision 🤦 .

> While making this change I added support on the response to identify a final result set (`next`) to avoid the user clicking "Show More" an extra time at the end, do you have a preference for supporting this in cursor-based approach? e.g. `paginate.before`

I think the look-ahead behavior that you implemented in the query is exactly right! Since cursor pagination is typically based on an exclusive filter, I think the way we would incorporate that as a conditional instruction for cursor-based pagination would look something like this:
```js
        pagination: {
            next: grantFollowersResult.length > limit ? grantFollowersResult[limit - 1].id : null,
        },
```
Essentially, we're instructing the client that 
1. If `pagination.next` is not `null`, then there are more results.
2. To continue paginating (in the same direction) from where the final result left off, provide the value of the final result from this current page, which will then be used as an exclusive filter on the next request.

Note that the above is pretty implementation-specific, as it relies on the concept that a higher `grant_followers.id` value implies a higher `grant_followers.created_at` value as well. If we were paginating by user name if alphabetical order, this wouldn't work. In fact, it could be argued that the more "correct" way to provide a cursor for follower recency would be to provide `pagination.next` as a `created_at` timestamp (I chose to go with `id` in the original spec because it avoids an edge case stemming from more than one row with the same `created_at` value). But that shouldn't really matter as long as the client is treating the cursor as an arbitrary value.

A more agnostic way to implement cursors that support multiple ordering schemes could be to provide the value of `grantFollowersResult[orderBy]` as the `pagination.next` cursor, but I think that gets a little dicey/brittle without having a clearer sense of what scenarios we'd need to support.

File: packages/client/src/components/Modals/GrantFollowers.vue [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+            email: user.email,
+            team: user.team.name,
+            dateFollowedText,
+          };
+        });
+    },
+    followersEmailText() {
+      return this.followers
+        .map((follower) => follower.user.email)
+        .join(',');
```
[the remark this answers is not in the document]
Comment (greg-adams) [reply packages/client/src/components/Modals/GrantFollowers.vue]: added

File: packages/client/src/components/Modals/GrantFollowers.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (greg-adams) [reply packages/client/src/components/Modals/GrantFollowers.vue]: updated to use luxon and locale-based output

**Comment by greg-adams** [comment]
No worries - I've updated to maintain use of `created_at`, keeping the `next` property to indicate an additional set

**Review by TylerHendrickson — APPROVED** [review]
@greg-adams Looks great!

**Pull request merged**